Feature/correct instruction documentation

martin-cs opened the pull request:
See https://github.com/diffblue/cbmc/pull/2031 for the discussion that started this.

Commit:
Expand the documentation of the goto-program instructions.

This clarifies a few sources of mistakes and misconceptions as well as
documenting some common practice.

tautschnig commented on src/goto-programs/goto_program.h:
```diff
@@ -134,17 +140,21 @@ class goto_programt
   /// - ATOMIC_BEGIN, ATOMIC_END:
   ///     When a thread executes ATOMIC_BEGIN, no thread other will be able to
   ///     execute any instruction until the same thread executes ATOMIC_END.
+  ///     Concurrency is not support by all analysis tools.
```
Nit pick: s/support/supported/

tautschnig commented on src/goto-programs/goto_program.h:
```diff
@@ -134,17 +140,21 @@ class goto_programt
   ///     Create a new thread and run the code of this function starting from
   ///     targets[0]. Quite often the instruction pointed by targets[0] will be
   ///     just a FUNCTION_CALL, followed by an END_THREAD.
+  ///     Concurrency is not support by all analysis tools.
```
As above.

tautschnig commented on src/goto-programs/goto_program.h:
```diff
@@ -134,17 +140,21 @@ class goto_programt
+  ///     Concurrency is not support by all analysis tools.
   /// - END_THREAD:
   ///     Terminate the calling thread.
+  ///     Concurrency is not support by all analysis tools.
```
As above.

tautschnig commented on src/goto-programs/goto_program.h:
```diff
@@ -106,14 +106,19 @@ class goto_programt
   ///     Introduces a symbol denoted by the field `code` (an instance of
   ///     code_declt), the life-time of which is bounded by a corresponding DEAD
-  ///     instruction.
+  ///     instruction.  Symbols must be DECL'd before they are used.
```
I'd refine this to "Non-static symbols must be ..."

Commit:
Correct the documentation of ASSERT : it does not alter control flow.

After a discussion on issue #2031, this was agreed to be the correct
semantics for the following reasons:

1. We want to check for conditions that do not result in termination.
For example, generation of Inf and NaN, reachability of locations and
unsigned overflow / underflow, error locations, exceptions or the use of
program approximations.  This is particularly important for binary
analysis and security.

2. Allowing traces with multiple assertion failures reduces the number of
solver calls and test cases generated, allowing minimisation without
complex program manipulation.

3. If ASSERT alters control flow then the truth of assertions can change
as more assertions are added, as the failures may get blocked.  Adding
more checks can result in less failures.

4. If ASSERT alters control flow, then it must be taken into account when
slicing, meaning that every variable in any assert must be part of the
slice.

5. ASSUME allows arbitrary blocking of execution, it is better than ASSERT
is orthogonal and:
ASSERT(X)
ASSUME(X)
is used for cases where assertion failure results in termination (such as
null pointer dereference or array bounds failures).

peterschrammel commented on src/goto-programs/goto_program.h:
```diff
@@ -106,14 +106,19 @@ class goto_programt
   /// - GOTO:
-  ///     if `guard` then goto `targets`
+  ///     goto `targets` if and only if `guard` is true
+  ///     If there are multiple targets, pick one non-deterministically.
```
Maybe add as a comment: Multiple targets are deprecated.

martin-cs commented:
All changes applied.
